File: packages/core/src/routing/outcomes.ts
/**
 * Application-recorded outcome memory — the input to quality-aware routing.
 *
 * Applications know quality better than any universal metric: a summarizer
 * can score faithfulness, an agent can score task completion. `recordOutcome`
 * ingests those signals as exponentially-weighted averages keyed by
 * (task, route). Strategies like "quality-first" read them; when no data
 * exists they degrade gracefully to cost/latency ordering.
 *
 * Memory is decaying, not permanent: quality and success EMAs drift toward
 * neutral (0.5) after `halfLifeMs` of silence, so stale verdicts cannot
 * outrank fresh ones forever. Deterministic EMA math only — no opaque ML.
 * `serialize()`/`load()` support cross-restart/cross-replica state sharing.
 */

export interface OutcomeEvent {
  /** Route id the request was served by. */
  routeId: string;
  /** Application task type, e.g. "summarize" or "support-ticket". Optional. */
  task?: string;
  /** Did the application consider the result usable? Default true. */
  success?: boolean;
  /**
   * Application-defined quality in [0,1] (human rating, eval harness,
   * downstream signal — anything). Never invented by the router itself.
   */
  quality?: number;
  latencyMs?: number;
  costUsd?: number;
}

export interface OutcomeStats {
  routeId: string;
  /** Task bucket ("" = untasked). */
  task: string;
  samples: number;
  /** EMA of success flags in [0,1]. */
  successRate: number;
  /** EMA of app-reported quality in [0,1]; undefined until reported. */
  avgQuality?: number;
  avgLatencyMs?: number;
  avgCostUsd?: number;
}

export interface OutcomeTrackerOptions {
  /** Smoothing factor for all EMAs (0 < alpha ≤ 1). Default 0.3. */
  alpha?: number;
  /** Half-life for decay toward neutral after silence in ms. 0 disables decay. */
  halfLifeMs?: number;
}

interface Bucket {
  samples: number;
  successRateEma: number;
  qualityEma?: number;
  latencyEma?: number;
  costEma?: number;
  lastAt: number;
}

const NEUTRAL = 0.5;

function emptyBucket(now: number): Bucket {
  return { samples: 0, successRateEma: 1, lastAt: now };
}

export class OutcomeTracker {
  private readonly buckets = new Map<string, Bucket>();
  private readonly alpha: number;
  private readonly halfLifeMs: number;

  constructor(opts: OutcomeTrackerOptions = {}) {
    this.alpha = Math.min(1, Math.max(0.001, opts.alpha ?? ALPHA_DEFAULT));
    this.halfLifeMs = opts.halfLifeMs ?? HALF_LIFE_DEFAULT;
  }

  record(event: OutcomeEvent): void {
    if (!/^[a-z0-9_.-]*$/i.test(event.routeId)) return; // defensive: never throw on telemetry
    if (event.quality !== undefined && !(event.quality >= 0 && event.quality <= 1)) return;
    const now = Date.now();
    const key = `${event.task ?? ""}|${event.routeId}`;
    const b = this.buckets.get(key) ?? emptyBucket(now);
    this.decay(b, now);
    b.samples++;
    b.successRateEma = ema(b.successRateEma, event.success === false ? 0 : 1, this.alpha);
    if (event.quality !== undefined) b.qualityEma = ema(b.qualityEma, event.quality, this.alpha);
    if (event.latencyMs !== undefined) b.latencyEma = ema(b.latencyEma, event.latencyMs, this.alpha);
    if (event.costUsd !== undefined) b.costEma = ema(b.costEma, event.costUsd, this.alpha);
    b.lastAt = now;
    this.buckets.set(key, b);
  }

  /** Observed quality EMA for a (task, route) pair; undefined without data. */
  quality(task: string | undefined, routeId: string): number | undefined {
    const b = this.buckets.get(`${task ?? ""}|${routeId}`);
    if (!b || b.qualityEma === undefined) return undefined;
    return this.decayed(b, Date.now())?.qualityEma;
  }

  snapshot(): OutcomeStats[] {
    const now = Date.now();
    const out: OutcomeStats[] = [];
    for (const [key, raw] of this.buckets) {
      const sep = key.indexOf("|");
      const task = key.slice(0, sep);
      const routeId = key.slice(sep + 1);
      const b = this.decayed(raw, now)!;
      out.push({
        routeId,
        task,
        samples: b.samples,
        successRate: round6(b.successRateEma),
        ...(b.qualityEma !== undefined ? { avgQuality: round6(b.qualityEma) } : {}),
        ...(b.latencyEma !== undefined ? { avgLatencyMs: round6(b.latencyEma) } : {}),
        ...(b.costEma !== undefined ? { avgCostUsd: round6(b.costEma) } : {}),
      });
    }
    return out;
  }

  // ------------------------------------------------------------- persistence

  serialize(now = Date.now()): unknown {
    const buckets: Array<[string, unknown]> = [];
    for (const [key, b] of this.buckets) {
      buckets.push([
        key,
        {
          samples: b.samples,
          successRateEma: b.successRateEma,
          qualityEma: b.qualityEma,
          latencyEma: b.latencyEma,
          costEma: b.costEma,
          lastAt: b.lastAt,
        },
      ]);
    }
    return { v: 1, savedAt: now, alpha: this.alpha, halfLifeMs: this.halfLifeMs, buckets };
  }

  load(data: unknown): boolean {
    if (typeof data !== "object" || data === null) return false;
    const d = data as { v?: number; buckets?: unknown };
    if (d.v !== 1 || !Array.isArray(d.buckets)) return false;
    try {
      for (const entry of d.buckets as Array<[string, Record<string, unknown>]>) {
        if (!Array.isArray(entry) || typeof entry[0] !== "string") continue;
        const r = entry[1];
        if (typeof r !== "object" || r === null) continue;
        const b: Bucket = {
          samples: numOr(r.samples, 0),
          successRateEma: clamp01(numOr(r.successRateEma, 1)),
          qualityEma: r.qualityEma === undefined || r.qualityEma === null ? undefined : clamp01(numOr(r.qualityEma, NEUTRAL)),
          latencyEma: numOrU(r.latencyEma),
          costEma: numOrU(r.costEma),
          lastAt: numOr(r.lastAt, Date.now()),
        };
        this.buckets.set(entry[0], b);
      }
      return true;
    } catch {
      return false;
    }
  }

  // ----------------------------------------------------------------- internals

  private decay(b: Bucket, now: number): void {
    if (this.halfLifeMs <= 0) return;
    const elapsed = now - b.lastAt;
    if (elapsed <= 0) return;
    const d = Math.pow(0.5, elapsed / this.halfLifeMs);
    // Success rate decays toward 50/50; quality toward neutral 0.5.
    b.successRateEma = NEUTRAL + (b.successRateEma - NEUTRAL) * d;
    if (b.qualityEma !== undefined) b.qualityEma = NEUTRAL + (b.qualityEma - NEUTRAL) * d;
  }

  private decayed(b: Bucket, now: number): Bucket | undefined {
    const copy: Bucket = { ...b };
    this.decay(copy, now);
    return copy;
  }
}

// ------------------------------------------------------------------ helpers

const ALPHA_DEFAULT = 0.3;
const HALF_LIFE_DEFAULT = 30 * 60_000;

function ema(prev: number | undefined, value: number, alpha: number): number {
  return prev === undefined ? value : prev * (1 - alpha) + value * alpha;
}

function round6(n: number): number {
  return Math.round(n * 1e6) / 1e6;
}

function clamp01(n: number): number {
  return Math.max(0, Math.min(1, n));
}

function numOr(v: unknown, fallback: number): number {
  return typeof v === "number" && Number.isFinite(v) ? v : fallback;
}

function numOrU(v: unknown): number | undefined {
  return typeof v === "number" && Number.isFinite(v) ? v : undefined;
}

File: packages/python/src/ai_router/routing/outcomes.py
"""Application-recorded outcome memory for quality-aware routing."""

from __future__ import annotations

import math
import time
from dataclasses import dataclass
from typing import Any

ALPHA_DEFAULT = 0.3
HALF_LIFE_DEFAULT = 30 * 60_000
NEUTRAL = 0.5


@dataclass
class OutcomeEvent:
    routeId: str
    task: str | None = None
    success: bool = True
    quality: float | None = None
    latencyMs: float | None = None
    costUsd: float | None = None


@dataclass
class OutcomeStats:
    routeId: str
    task: str
    samples: int
    successRate: float
    avgQuality: float | None = None
    avgLatencyMs: float | None = None
    avgCostUsd: float | None = None


@dataclass
class Bucket:
    samples: int
    successRateEma: float
    lastAt: float
    qualityEma: float | None = None
    latencyEma: float | None = None
    costEma: float | None = None


def _ema(prev: float | None, next_val: float, alpha: float) -> float:
    if prev is None:
        return next_val
    return prev + alpha * (next_val - prev)


def _round6(n: float) -> float:
    return round(n, 6)


class OutcomeTracker:
    def __init__(
        self,
        alpha: float = ALPHA_DEFAULT,
        halfLifeMs: int = HALF_LIFE_DEFAULT,
    ) -> None:
        self.alpha = min(1.0, max(0.001, alpha))
        self.halfLifeMs = halfLifeMs
        self._buckets: dict[str, Bucket] = {}

    def _decay(self, b: Bucket, now: float) -> None:
        if self.halfLifeMs <= 0:
            return
        elapsed = now - b.lastAt
        if elapsed <= 0:
            return
        d = math.pow(0.5, elapsed / self.halfLifeMs)
        b.successRateEma = NEUTRAL + (b.successRateEma - NEUTRAL) * d
        if b.qualityEma is not None:
            b.qualityEma = NEUTRAL + (b.qualityEma - NEUTRAL) * d

    def record(self, event: OutcomeEvent | dict[str, Any]) -> None:
        if isinstance(event, dict):
            ev = OutcomeEvent(**event)
        else:
            ev = event

        if ev.quality is not None and not (0.0 <= ev.quality <= 1.0):
            return

        now = time.time() * 1000.0
        key = f"{ev.task or ''}|{ev.routeId}"
        b = self._buckets.get(key)
        if not b:
            b = Bucket(samples=0, successRateEma=1.0, lastAt=now)
            self._buckets[key] = b

        self._decay(b, now)
        b.samples += 1
        b.successRateEma = _ema(b.successRateEma, 0.0 if ev.success is False else 1.0, self.alpha)
        if ev.quality is not None:
            b.qualityEma = _ema(b.qualityEma, ev.quality, self.alpha)
        if ev.latencyMs is not None:
            b.latencyEma = _ema(b.latencyEma, ev.latencyMs, self.alpha)
        if ev.costUsd is not None:
            b.costEma = _ema(b.costEma, ev.costUsd, self.alpha)
        b.lastAt = now

    def quality(self, task: str | None, routeId: str) -> float | None:
        key = f"{task or ''}|{routeId}"
        b = self._buckets.get(key)
        if not b or b.qualityEma is None:
            return None
        now = time.time() * 1000.0
        self._decay(b, now)
        return b.qualityEma

    def snapshot(self) -> list[OutcomeStats]:
        now = time.time() * 1000.0
        out: list[OutcomeStats] = []
        for key, b in self._buckets.items():
            sep = key.find("|")
            task = key[:sep]
            route_id = key[sep + 1 :]
            self._decay(b, now)
            out.append(
                OutcomeStats(
                    routeId=route_id,
                    task=task,
                    samples=b.samples,
                    successRate=_round6(b.successRateEma),
                    avgQuality=_round6(b.qualityEma) if b.qualityEma is not None else None,
                    avgLatencyMs=_round6(b.latencyEma) if b.latencyEma is not None else None,
                    avgCostUsd=_round6(b.costEma) if b.costEma is not None else None,
                )
            )
        return out
